clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -358,15 +360,23 @@ public void wakeup() {
     @Override
     public void close() {
         List<String> connections = new ArrayList<>(channels.keySet());
-        for (String id : connections)
-            close(id);
         try {
-            this.nioSelector.close();
-        } catch (IOException | SecurityException e) {
-            log.error("Exception closing nioSelector:", e);
+            for (String id : connections)
+                close(id);
+        } finally {
+            // If there is any exception thrown in close(id), we should still be able
+            // to close the remaining objects, especially the sensors because keeping
+            // the sensors may lead to failure to start up the ReplicaFetcherThread if
+            // the old sensors with the same names has not yet been cleaned up.
+            AtomicReference<Throwable> firstException = new AtomicReference<>();
+            ClientUtils.closeQuietly(nioSelector, "nioSelector", firstException);
+            ClientUtils.closeQuietly(sensors, "sensors", firstException);
+            ClientUtils.closeQuietly(channelBuilder, "channelBuilder", firstException);
+            Throwable exception = firstException.get();
+            if (exception instanceof RuntimeException) {
```

Comment:
Is the change in behavior intentional? `SecurityException`, which was previously caught and logged, is an instance of `RuntimeException`.

Reply:
Good catch. I miss the part that SecurityException is a RuntimeException. Given that we will only throw runtime exception her, my intention was to check whether there is any RuntimeException and throw it. I will exclude `SecurityException` in this case.